kolibri/core/assets/src/views/k-textbox/index.vue

```diff
@@ -62,18 +64,16 @@
         default: false,
       },
       /**
-       * Whether or not input is invalid
+       * A list of required validations. Validations are objects that consists of a validation function and an invalid text.
        */
-      invalid: {
-        type: Boolean,
-        default: false,
-      },
-      /**
-       * Text displayed if input is invalid
-       */
-      invalidText: {
-        type: String,
-        required: false,
+      validations: {
+        type: Array,
+        default: () => [],
+        validator(value) {
+          return value.every(
+            item => item.hasOwnProperty('validator') && item.hasOwnProperty('invalidText')
+          );
+        },
```

Comment:
Sorry for the block of text below. Form validation is a notoriously tricky problem and can be difficult to do in a declarative, extensible way.

Delegating the responsibility of validation to the textbox by passing in validator functions might give too much responsibility and local state to the textbox component. In particular it leads to imperative-style coding like in the [create-user modal](https://github.com/learningequality/kolibri/pull/2005/commits/736dd41582b415a7878e2e7a7ce796b99e9418e4#diff-3add6dedbfbaa28f7d2c673b70865ffaR188), where you have to command the textbox to perform a validation with `textboxRef.validate()` and query the validity state with `textboxRef.isValid()`.

Another risk is that this bakes some assumptions about how we do validation in forms - in particular, waiting until de-focus to validate - into the textbox component. That's good a good thing until we come across one exception to the rule where we want to validate on keypress.

On the other hand, one difficulty with moving validation state to the parent form component is that `isValid` for a field does depend on keyboard focus state changes. A way around that might be to not use `v-model` (which updates the parent data on every keypress) and instead listen for [`change` events](https://www.w3.org/WAI/UA/TS/html401/cp0102/0102-ONCHANGE-INPUT.html) (which should only get triggered when the textbox is blurred). This would decouple the "current value visible in the box" from the "value that determines what error message should be displayed".

Quoting from the [forms style guide](https://docs.google.com/document/d/1rrq8fovV5wBtUKv65wfRqXxP4SfnV30anUZHfzCVmN4/edit):

> * Error text should only be shown after the user interacts with the field **and** loses focus on it. 
> * Error text should disappear once the user alters anything within an errored field **and** loses focus on it.
> * If there is a field that is clearly marked with an error on the client side, the submit button for the form should be disabled until the client side errors on each affected field are gone.

So I _think_ as long as we wait to update the parent form `data` until de-focus, all of the validation state can simply be computed properties of those parent `data` values and those would reactively control the field messaging and submission button.

Over time we might see some common patterns (e.g. managing pairs of validation functions and message strings) repeated over and over at the form level. When that happens, we could possibly look at creating a form mixin with some helpers?

Reply:
Thanks for the feedback. Will take all that this into account.

Reply:
FYI, I realized that `change` events should also be triggered on `enter`. This will ensure the new value is propagated up before e.g. a form is submitted.